aten/src/ATen/native/RNN.cpp

```diff
@@ -0,0 +1,700 @@
+  using unstacked_output_type = LayerOutput<std::vector<Tensor>, hidden_type>;
+
+  FullLayer(Cell<hidden_type>& cell)
+    : cell_(cell) {};
+
+  unstacked_output_type operator()(std::vector<Tensor> step_inputs, const hidden_type& input_hidden, const CellParams& params) const {
+    std::vector<Tensor> step_outputs;
+    auto hidden = input_hidden;
+    for (size_t i = 0; i < step_inputs.size(); i++) {
+      hidden = cell_(step_inputs[i], hidden, params);
+      step_outputs.push_back(hidden_as_output(hidden));
+    }
+    return {step_outputs, hidden};
+  }
+
+  output_type operator()(const Tensor& inputs, const hidden_type& input_hidden, const CellParams& params) const override {
+    auto unstacked_output = (*this)(inputs.unbind(0), input_hidden, params);
+    return {at::stack(unstacked_output.outputs, 0), unstacked_output.final_hidden};
+  }
+
+  Cell<hidden_type>& cell_;
+};
+
+template<typename dir_hidden_type>
+struct FullBidirectionalLayer : Layer<Tensor, pair_of<dir_hidden_type>, pair_of<CellParams>> {
+  using hidden_type = pair_of<dir_hidden_type>;
+  using param_type = pair_of<CellParams>;
+  using output_type = typename Layer<Tensor, hidden_type, param_type>::output_type;
+
+  FullBidirectionalLayer(Cell<dir_hidden_type>& cell)
+    : layer_(cell) {};
+
+  output_type operator()(const Tensor& input, const hidden_type& input_hidden, const param_type& params) const override {
+    auto step_inputs = input.unbind(0);
+    auto fw_result = layer_(step_inputs, input_hidden.first, params.first);
+    auto fw_output = at::stack(fw_result.outputs, 0);
+
+    auto rev_step_inputs = reverse(std::move(step_inputs));
+    auto rev_result = layer_(rev_step_inputs, input_hidden.second, params.second);
+    std::reverse(rev_result.outputs.begin(), rev_result.outputs.end());
+    auto rev_output = at::stack(rev_result.outputs, 0);
+
+    return {at::cat({fw_output, rev_output}, fw_output.dim() - 1),
+            std::make_pair(fw_result.final_hidden, rev_result.final_hidden)};
+  }
+
+  std::vector<Tensor> reverse(std::vector<Tensor>&& x) const {
+    std::reverse(x.begin(), x.end());
+    return x;
+  }
+
+  FullLayer<dir_hidden_type> layer_;
+};
+
+template<typename hidden_type>
+struct PackedLayer : Layer<PackedSequence, hidden_type, CellParams> {
+  using output_type = typename Layer<PackedSequence, hidden_type, CellParams>::output_type;
+
+  PackedLayer(Cell<hidden_type>& cell)
+    : cell_(cell) {};
+
+  output_type operator()(const PackedSequence& input, const hidden_type& input_hidden, const CellParams& params) const override {
+    std::vector<at::Tensor> step_outputs;
+    std::vector<hidden_type> hiddens;
+    int64_t input_offset = 0;
+    int64_t num_steps = input.batch_sizes.size(0);
+    int64_t* batch_sizes = input.batch_sizes.data<int64_t>();
+    int64_t last_batch_size = batch_sizes[0];
+
+    // TODO: use split with sizes?
+    auto hidden = input_hidden;
```

Comment:
So... do you know what this code is doing? Would you mind adding a comment explaining it? It is breaking my brain.